### PR TITLE
Refactor/aniso pair structs

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR updates all Evaluators used with AnisoPairPotential to use a struct for managing their parameters. All such structs are equipped with the ability to load their data into shared memory. This change is enabled by updates to GlobalArray that allow client code to force it to use shared memory.

## Motivation and Context

These changes allow much more efficient access to pair potential parameters if they are more complex than just a few constants. For example, if any such potential involved a table lookup, the table could now be loaded into shared memory.

## How Has This Been Tested?

These changes are tested on a separate development branch implementing a new anisotropic potential Evaluator.

## Change log

```
Allow client code to force GlobalArrays to use shared memory and update all anisotropic potential Evaluators to use structs that can leverage this feature if needed.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).
- [x] (Glotzer group members): I am a member of the [hoomd-contributors team](https://github.com/orgs/glotzerlab/teams/hoomd-contributors/members).

### CODE REVIEW

File: hoomd/md/AnisoPotentialPairGPU.cuh [hunk 1]
```diff
@@ -398,8 +410,37 @@ struct AnisoPairForceComputeKernel
...
 
             static unsigned int max_block_size = UINT_MAX;
+            cudaFuncAttributes attr;
             if (max_block_size == UINT_MAX)
-                max_block_size = aniso_get_max_block_size(gpu_compute_pair_aniso_forces_kernel<evaluator, shift_mode, compute_virial, tpp>);
+                {
+                cudaFuncGetAttributes(&attr, gpu_compute_pair_aniso_forces_kernel<evaluator, shift_mode, compute_virial, tpp>);
+                int max_threads = attr.maxThreadsPerBlock;
+                // number of threads has to be multiple of warp size
+                max_block_size = max_threads - max_threads % gpu_aniso_pair_force_max_tpp;;
```
Comment (tcmoore3) [inline hoomd/md/AnisoPotentialPairGPU.cuh]: Extra semicolon

Comment (vyasr) [reply hoomd/md/AnisoPotentialPairGPU.cuh]: Thanks, fixed.

**Comment by vyasr** [comment]
@tcmoore3 there are new tests but they are being done on a separate, active line of development. The changes in this PR are the ones that are most likely to affect other existing code paths, so I chose to make a separate pull request to avoid too much code divergence. Further tests will be merged in at a later date.

**Pull request merged**